### DISCUSSION TITLE
Weird concurrent modification

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#83] [Off topic]
Hi, I was receiving a weird concurrent modification error on Dart Web with this code: I solved it by cloning the list/set before iterating it. Can you think a reason why that error happens if there is not concurrent possibility (all three methods are sync)?

```dart
class AppKeyEventManager {
  final String name;
  final _listeners = <KeyEventListener>{};

  AppKeyEventManager(this.name);

  bool addListener(KeyEventListener listener) => _listeners.add(listener);

  bool removeListener(KeyEventListener listener) => _listeners.remove(listener);

  void notifyListeners(KeyboardEvent event) {
    var preventDefault = false;
    var stopPropagation = false;

    // Cloning into list in order to prevent weird error concurrent modification
    final listeners = [..._listeners];

    for (final listener in listeners) {
      final cleanUp = listener(event);

      switch (cleanUp) {
        case KeyEventCleanUp.preventDefault:
          preventDefault = true;
        case KeyEventCleanUp.stopPropagation:
          stopPropagation = true;
        case KeyEventCleanUp.full:
          preventDefault = true;
          stopPropagation = true;
        default:
      }
    }

    debug(
      'cleanup -- $name -- preventDefault: $preventDefault, stopPropagation: $stopPropagation',
      active: cc.read(appDebugCapsule).keyEvents.cleanUp,
    );

    if (preventDefault) {
      event.preventDefault();
    }

    if (stopPropagation) {
      event.stopPropagation();
    }
  }
}
```

**Comment by busslina** [comment]
The only way would be if any listener call `addListener` or `removeListener`. Checking it

**Reply by GregoryConrad** [reply]
Was just about to say that. `listeners` would only be modified within the for loop if one of the listeners adds/removes listeners. Maybe you remove a listener in the listener itself in order to prevent a leak or something (which in turn causes that exception)?

**Reply by busslina** [reply]
Yeah, I was pressing escape (27) which was replacing the current route with the previous (Dart React). This operation I thought it is not done synchronously (to un/mount a React component) inside my listener, but I think it is. 

Also print statements were not printed in order (some of them were executed before the error but printed after it) and was causing me some troubles.

Final solution: using `synchronized`'s `Lock` on top of that three methods.

;)
